exp/operator/controllers/phases.go

```diff
@@ -0,0 +1,417 @@
+	options            repository.ComponentsOptions
+	providerConfig     configclient.Provider
+	configClient       configclient.Client
+	components         repository.Components
+	clusterctlProvider *clusterctlv1.Provider
+}
+
+type reconcilePhaseFn func(context.Context, genericprovider.GenericProvider) (reconcile.Result, error)
+
+type PhaseError struct {
+	Reason   string
+	Type     clusterv1.ConditionType
+	Severity clusterv1.ConditionSeverity
+	Err      error
+}
+
+func (s *PhaseError) Error() string {
+	return s.Err.Error()
+}
+
+func NewPreFlightError(err error, reason string) error {
+	if err == nil {
+		return nil
+	}
+	return &PhaseError{
+		Err:      err,
+		Type:     operatorv1.PreflightCheckCondition,
+		Reason:   reason,
+		Severity: v1alpha4.ConditionSeverityWarning,
+	}
+}
+
+func NewCertManagerReadyError(err error, reason string) error {
+	if err == nil {
+		return nil
+	}
+	return &PhaseError{
+		Err:      err,
+		Type:     operatorv1.CertManagerReadyCondition,
+		Reason:   reason,
+		Severity: v1alpha4.ConditionSeverityError,
+	}
+}
+
+func NewInstallReadyError(err error, reason string) error {
+	if err == nil {
+		return nil
+	}
+	return &PhaseError{
+		Err:      err,
+		Type:     operatorv1.ProviderInstalledCondition,
+		Reason:   reason,
+		Severity: v1alpha4.ConditionSeverityError,
+	}
+}
+
+func newReconcilePhases(c client.Client, certManagerInstaller SingletonInstaller) *reconciler {
+	return &reconciler{
+		ctrlClient:           c,
+		certManagerInstaller: certManagerInstaller,
+		clusterctlProvider:   &clusterctlv1.Provider{},
+	}
+}
+
+func (s *reconciler) load(ctx context.Context, provider genericprovider.GenericProvider) (reconcile.Result, error) {
+	reader, err := s.secretReader(ctx, provider)
+	if err != nil {
+		return reconcile.Result{}, NewPreFlightError(err, "failed to load the secret reader")
+	}
+
+	s.configClient, err = configclient.New("", configclient.InjectReader(reader))
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	s.providerConfig, err = s.configClient.Providers().Get(provider.GetName(), util.ClusterctlProviderType(provider))
+	if err != nil {
+		return reconcile.Result{}, NewPreFlightError(err, operatorv1.UnknownProviderReason)
+	}
+
+	spec := provider.GetSpec()
+
+	if spec.FetchConfig != nil && spec.FetchConfig.Selector != nil {
+		s.repo, err = s.configmapRepository(ctx, provider)
+	} else {
+		s.repo, err = repository.NewGitHubRepository(s.providerConfig, s.configClient.Variables())
+	}
+	if err != nil {
+		return reconcile.Result{}, NewPreFlightError(err, "failed to load the repository")
+	}
+
+	s.options = repository.ComponentsOptions{
+		TargetNamespace:     provider.GetNamespace(),
+		SkipTemplateProcess: false,
+		Version:             s.repo.DefaultVersion(),
+	}
+	if spec.Version != nil {
+		s.options.Version = *spec.Version
+	}
+
+	err = s.validateRepoCAPIVersion(provider)
+	return reconcile.Result{}, NewPreFlightError(err, operatorv1.CAPIVersionIncompatibilityReason)
```

Comment:
We don't necessarily know that `NewPreFlightError` will check for a non-nil error while reading this part of the code, might be better to have an explicit `if err != nil` check here to make this function easier to follow